packages/ruleset/src/utils/get-composite-schema-attribute.js

```diff
@@ -0,0 +1,25 @@
+const checkCompositeSchemaForConstraint = require('./check-composite-schema-for-constraint');
+
+/**
+ * Retrieves the value of "schema"'s attribute named "attrName" either directly from "schema"
+ * or from one of its composition "children".
+ *
+ * This function can be used in scenarios where you need to retrieve a particular schema's
+ * field (e.g. the pattern field), but the field might exist either in "schema" itself,
+ * or within one or more allOf/anyOf/oneOf list element schemas.
+ * @param {*} schema the schema to retrieve the attribute from
+ * @param {*} attrName the name of the attribute to retrieve
+ * @returns the value of the attribute or undefined if not present
+ */
+function getCompositeSchemaAttribute(schema, attrName) {
+  let value = undefined;
+  const foundConstraint = checkCompositeSchemaForConstraint(schema, s => {
+    if (attrName in s && s[attrName] !== undefined && s[attrName] !== null) {
+      value = s[attrName];
+      return true;
+    }
+  });
+  return foundConstraint ? value : undefined;
```

Comment:
Seems this is probably equivalent 🙂 

```suggestion
  return value;
```

But the code is definitely good and clear as is

Reply:
I initially thought so too, but that is actually not correct because it's possible that we were able to save off the value of the attribute from an individual anyOf/oneOf list element if we happened to find it there, but yet we would still get back a false from checkComposedSchemaForConstraint() if for example the attribute was not present in all elements of an anyOf or oneOf list.   So in that scenario, we'd want to return undefined for the attribute value.

Reply:
Ohh right, I didn't even think about that 🤦‍♂️

Thanks for the explanation!